core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -797,6 +819,43 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     }
   }
 
+  /**
+   * Check and delete specified event log according to the max log age defined by the user.
+   */
+  private def checkAndCleanLog(logPath: String): Unit = Utils.tryLog {
+    val maxTime = clock.getTimeMillis() - conf.get(MAX_LOG_AGE_S) * 1000
+    val expiredLog = Try {
+      val log = listing.read(classOf[LogInfo], logPath)
+      if (log.lastProcessed < maxTime) Some(log) else None
```

Comment:
This is wrong. The attempt may be expired but the `lastProcessed` time may be recent. So you always need to check the app data itself for expiration.

`lastProcessed` is only used for log files that never record the app ID (e.g. the driver crashed soon after creating the file).

Reply:
BTW you should really add a unit test. This is basically `checkForLogs()` checking for the age of an app and deleting it if needed, without having to call `cleanLogs`.

And that reminds me, you shouldn't delete anything if the cleaner is disabled.

(After you fix this there's a chance that the cleaner task becomes redundant. But I'd look at cleaning that up separately.)

Reply:
Sorry for late response. I have addressed the comments you mentioned and added unit test.
Thanks very much.